OSGi: Restrict Import-Package for scala.*

lefou opened the pull request:
Restrict the Import-Package OSGi manifest entry to match only binary compatible scala.\* packages. This is necessary, because the Scala OSGi versions do not match the OSGi semantic versioning policy and are not binary compatible between their OSGi minor versions. This means that e.g. the scala-compiler-2.11 bundle will never work if it imports a scala.\* package from 2.10 or 2.12. Thus this patch enforces a tight version range, only accepting versions within the same minor version. E.g. Applied to version 2.11.1 the range would expand to "[2.11,2.12)".

This PR also enforces a Java 6 execution environment for the bundles.

Suggested reviewer: @jsuereth

lefou committed:
Restrict Import-Package for scala.*

Restrict the Import-Package OSGi manifest entry to match only binary compatible scala.* packages. This is necessary, because the Scala OSGi versions do not match the OSGi semantic versioning policy and are not binary compatible between their OSGi minor versions. This means that e.g. the scala-compiler-2.11 bundle will never work if it imports a scala.* package from 2.10 or 2.12. Thus this patch enforces a tight version range, only accepting versions within the same minor version. E.g. Applied to version 2.11.1 the range would expand to "[2.11,2.12)".

Additionally, a Java 6 execution environment will be enforced for OSGi.

All module bundles need to have their own version numbers. This was not the case before this commit. Although these module bundles were only build for testing, they are needed to be correct, now, that we have stricter package import constraints and some test cases rely on them.

lefou commented:
Can somebody help me understand the issue here?

retronym commented:
The build failure in trying to Scala IDE with the new version of the compiler.

```
[INFO] Resolving dependencies of MavenProject: org.scala-refactoring:org.scala-refactoring.library:0.6.2-SNAPSHOT @ /localhome/jenkins/a/workspace/pr-scala-integrate-ide/target/scala-refactoring/org.scala-refactoring.library/pom.xml
[INFO] Cannot complete the request.  Generating details.
[INFO] Cannot complete the request.  Generating details.
[INFO] {osgi.ws=gtk, osgi.os=linux, osgi.arch=x86_64, org.eclipse.update.install.features=true}
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: org.scala-refactoring.library 0.6.2.qualifier
[ERROR]   Missing requirement: org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb requires 'package scala.util.parsing.json [2.11.0,2.12.0)' but it could not be found
[ERROR]   Cannot satisfy dependency: org.scala-refactoring.library 0.6.2.qualifier depends on: package scala.tools.nsc 0.0.0
[ERROR] 
[ERROR] Internal error: java.lang.RuntimeException: "No solution found because the problem is unsatisfiable.": ["Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.util.parsing.json [2.11.0,2.12.0).", "Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.xml [2.11.0,2.12.0).", "Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.xml.dtd [2.11.0,2.12.0).", "Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.xml.parsing [2.11.0,2.12.0).", "No solution found because the problem is unsatisfiable."] -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: "No solution found because the problem is unsatisfiable.": ["Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.util.parsing.json [2.11.0,2.12.0).", "Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.xml [2.11.0,2.12.0).", "Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.xml.dtd [2.11.0,2.12.0).", "Unable to satisfy dependency from org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb to package scala.xml.parsing [2.11.0,2.12.0).", "No solution found because the problem is unsatisfiable."]
```

Perhaps one of @skyluc @huitseeker @dragos can help out here.

lefou commented:
Thank you, @retronym. Now, I found the output, from where you cut the snippet above.

Though, I can't fully understand the issue. The message seems to come from a running OSGi container trying to resolve the bundle `org.scala-refactoring.library 0.6.2.qualifier`. It imports the package `scala.tools.nsc` without any version constraint, which should be provided by bundle `org.scala-lang.scala-compiler 2.11.1.v20140424-234447-814fcb3-SNAPSHOT-4425ce34eb`. But the scala-compiler bundle can't be resolved because of missing imports like `package scala.util.parsing.json [2.11.0,2.12.0)` and `package scala.xml [2.11.0,2.12.0)` and others.

Are the bundles containing these packages (`org.scala-lang.modules.scala-parser-combinators`, `org.scala-lang.modules.scala-xml`) missing in the container?

I can't tell and unfortunately, I don't know how to reproduce it here. How to set it up?  

Of course, this is not an issue with this PR per se. Providing a fix by only changing this PR, without also touching the Scala IDE build setups, seems to be very hard for me.

skyluc commented:
This package (and others) has been extracted into modules in 2.11, and the resulting jars have different version numbers.
The current version numbers are defined in https://github.com/scala/scala/blob/master/versions.properties

lefou commented:
@skyluc That was my first idea, too, after I saw the error message, but then, I examined the output of the `ant osgi.done` build step and found this:

```
$ zcat ./build/osgi/org.scala-lang.modules.scala-parser-combinators.jar | head
Manifest-Version: 1.0
Bnd-LastModified: 1398423086877
Bundle-ManifestVersion: 2
Bundle-Name: Scala Parser Combinators Library
Bundle-RequiredExecutionEnvironment: JavaSE-1.6
Bundle-SymbolicName: org.scala-lang.modules.scala-parser-combinators
Bundle-Version: 2.11.1.v20140424-163056-814fcb3388
Created-By: 1.7.0_55 (Oracle Corporation)
Export-Package: scala.util.parsing.combinator.lexical;uses:="scala.runti
 me,scala,scala.collection,scala.util.parsing.combinator,scala.util,scal
```

The built bundle has clearly not the version 1.0.1, which I had expected, but 2.11.1.v20140424-163056-814fcb3388.

So, where came these bundles from, when they are not the ones built with `ant osgi.done`?

skyluc commented:
Scala IDE (and scala-refactoring) references the modules using the binary compatible artifact id, and the 'real' version: `scala-xml_2.11` version `1.0.1`.

They are generated in a different process, and made available through maven repositories.

I'm not sure how you get the jars with same version as Scala, but they are definitively not the ones that are shipped with a Scala distribution.

lefou commented:
I reviewed the package imports and came to the conclusion, that only scala-compiler makes use of the modules. I added proper import version ranges for them.

What's the version policy of the scala modules? Are they guaranteed to be binary compatible between micro versions? Are micro versions backward compatible. For now I enforced a version range of `[===,=+)` which means, the same major and minor version and at least the same micro version or greater. So, for e.g. scala.xml 1.0.1 the import version range will be `[1.0.1,1.1)`.

lefou commented:
PLS REBUILD/pr-scala@50ce33f

scala-jenkins commented:
(kitty-note-to-self: ignore 41446854)
:cat: Roger! Rebuilding pr-scala for 50ce33f8. :rotating_light:

lefou commented:
PLS REBUILD/pr-scala@32db35f

scala-jenkins commented:
(kitty-note-to-self: ignore 41505969)
:cat: Roger! Rebuilding pr-scala for 32db35f3. :rotating_light:

lefou commented:
I'm not exactly sure, what's the proceeding now? Shall I squash the failed commits? For now, I will wait for review feedback...

lefou commented:
@retronym Am I supposed to squash commit and push force this PR, before any review happens? The first commits have test failures which were only fixed by the last commit.

retronym commented:
All commits in the history should be green, so please squash or reorder as needed. Hopefully then @jsuereth and @skyluc can take a close look at this.

lefou commented:
@jsuereth, @skyluc  and @dragos would you please review this PR?

huitseeker commented on src/build/bnd/scala-actors.bnd:
```diff
@@ -3,3 +3,5 @@ Bundle-SymbolicName: org.scala-lang.scala-actors
 Bundle-Version: ${ver}
 Export-Package: *;version=${ver}
+Import-Package: scala.*;version="${range;[==,=+);${ver}}",*
+Bundle-RequiredExecutionEnvironment: JavaSE-1.6
```
While support for `JavaSE-1.7` has been somewhat experimental for some time, it's actually supported here. How about also listing it along 1.6 ?

lefou replied:
As far as I know `JavaSE-1.7` is compatible to `JavaSE-1.6` and I never saw an OSGi container running on a Java 7 JRE rejecting any `JavaSE-1.6` bundle. So it is redundant, IMHO. Besides that, it would be perfectly valid to add it, as it is only required, that at least one listed environment is satisfied.

huitseeker replied:
Yes, indeed. But some OSGI containers running on Java 7 do routinely spew out warnings when required a Java SE 1.6 bundle (to wit: Eclipse). This might give the wrong impression to people, So I'd be in favor of running a `sed -re 's/(Bundle-RequiredExecutionEnvironment: JavaSE-1\.6)/\1, JavaSE-1.7/'` on this PR.

huitseeker commented on src/build/bnd/scala-compiler.bnd:
```diff
@@ -5,4 +5,9 @@ Bundle-Version: ${ver}
 Export-Package: *;version=${ver}
 Import-Package: jline.*;resolution:=optional, \
                 org.apache.tools.ant.*;resolution:=optional, \
+                scala.util.parsing.*;version="${range;[===,=+);@PARSER_COMBINATORS_VERSION@}", \
```
I'm not sure this should declare a dependency on parser-combinators and xml (at most, the xml dependency is only needed for scaladoc). Pinging @adriaanm 

But in any case, the dependency, if it exists at all, should be strict. To wit:
- at the moment, scala 2.11.0-RC1 .. 2.11.0-RC5, 2.11.0 have depended on the same, untouched verison of parser-combinators : 1.0.1.
- but one can imagine some change that would require releasing a binary-incompatible 1.1.1 Or for that matter a binary-incompatible 1.0.2. (to my knowledge there's no extant agreement on semantic versioning for combinators)
- in that case, the scala compiler may want to require a very strict version of combinators/xml.

lefou replied:
Those dependencies were in the generated bundles, even before I added them to refine their versions. If fact, the `*` will only expand if bnd finds any actual imports.

`scala.util.parsing.*`: If there is no guarantee, that any other version may be compatible, I will change the version to `[===,==+)`. Except, you tell me, that even the qualifier might bring an incompatible version, in which case a `[====,====]` might be required.

`scala.xml`: Do you mean, they are only referenced in the scaladoc? In that case, I could explicitly exclude them. Else, a `resolution:=optional` might be a good option. What about compatible versions? Same as in `scala.util.parsing`?

huitseeker replied:
I'm sure you can add the optional for xml, and 99% sure you can for parsing as well. For the strict versioning, it's a matter of choosing a convention rather than technical dependency (so far combinators & xml are maintenance-only, so there's just one version to speak of) On that matter I'll have to defer to @adriaanm, @retronym, or @gkossakowski.

huitseeker replied:
OK, so after some checking with @skyluc :
- the dependency  on scala-xml is only from scaladoc, so the resolve:optional is pertinent
- the dependency on scala-parser-combinators is non-existent, the poms should be cleaned
- `test/disabled/presentation/akka/src/akka/config/ConfigParser.scala` should go away or be refactored to avoid the dependency on combinators

practical consequences for this PR:
- strict versioning (`[===, ===]`) AND `resolution:= optional` for scala -xml seems the best option
- you can remove the scala-parser-combinators dependency

/cc @phaller with whom I talked about this

lefou committed:
Add JavaSE-1.7 to set of required execution environments.

Without it, Equinox on Java 7 might spit out some warnings, which might lead to the wrong impression, that the Scala bundles are not well supported for newer platforms.

lefou committed:
Refined package imports for scala modules.

Made scala.xml.* import optional and removed scala.util.parsing.json import.

adriaanm commented on src/build/bnd/scala-compiler.bnd:
```diff
@@ -5,4 +5,8 @@ Bundle-Version: ${ver}
 Export-Package: *;version=${ver}
 Import-Package: jline.*;resolution:=optional, \
                 org.apache.tools.ant.*;resolution:=optional, \
+                !scala.util.parsing.*, \
```
what does this do / why is it done?

lefou replied:
This removes all package imports for packages including and below `scala.util.parsing`, even if bnd / maven-bundle-plugin thinks, it is required. I added this exclusion rule based on @huitseeker last comment (after he talked with @phaller about that):

> practical consequences for this PR:
> - strict versioning ([===, ===]) AND resolution:= optional for scala -xml seems the best option
> - you can remove the scala-parser-combinators dependency

Although, I have to admit, that I pretty much prefer to leave it there but mark it optional, like I did with `scala.xml` packages. In general, I trust bnd's bytecode analysis, which means, there must be a real dependency to `scala.util.parsing` somewhere, and without knowing details, I rather would not prohibit the import of it.

In the current published scala-compiler bundle, there **is** a non-optional import for `scala.util.parsing`, but without any version constraints.

It's your call.

retronym replied:
![image](https://cloud.githubusercontent.com/assets/65551/2928803/ac6801ce-d781-11e3-8697-2507e89ae051.png)

retronym replied:
Optional, ala xml, seems like the best option here.

adriaanm replied:
If there's a real dependency, and we forbid importing the package, will this yield a NoClassDef error when e.g., JSONObject is required at run-time? (More generally, would this mean scaladoc cannot run in OSGi?)

lefou replied:
@adriaanm Without knowing the details (which class does what), forbidding a real dependency might yield a NoClassDefError or a ClassNotFoundException, yes.

I will push a new commit re-adding that import, but optionally.

lefou committed:
Re-added the scala.util.parsing package import, with optional resolution.

retronym commented:
Looks like this PR is getting very close! Thanks for your persistence. But I have to ask you to jump one more hurdle: could you please resubmit against the 2.11.x branch? We are switching our branching structure and I am retiring the master branch.
